src/app/static/src/app/components/genericChart/GenericChart.vue

```diff
@@ -252,7 +295,46 @@
                 return !this.chartData ||
                     !Object.values(this.chartData).some(e => e.length);
 
-            }
+            },
+            allDistinctColumnValues() {
+                const subplots = this.chartMetadata.subplots;
+                if (subplots && this.chartData) {
+                    return this.chartData["data"].reduce((distinct: string[], row: any) => {
+                        const distinctVal = row[subplots.distinctColumn];
+                        return distinct.includes(distinctVal) ? distinct : [...distinct, distinctVal]
+                    }, []);
+                } else {
+                    return null;
+                }
+            },
+            pageDistinctColumnValues() {
+                if (this.chartData && this.chartMetadata.subplots && this.allDistinctColumnValues) {
+                    const subplotsPerPage = this.chartMetadata.subplots.subplotsPerPage;
+                    const start = (this.currentPage - 1) * subplotsPerPage;
+                    return this.allDistinctColumnValues.slice(start, start + subplotsPerPage);
+                } else {
+                    return null;
+                }
+            },
+            totalPages() {
+                return this.allDistinctColumnValues ? Math.ceil(this.allDistinctColumnValues.length / this.chartMetadata.subplots!.subplotsPerPage) : 1;
+            },
+            chartDataPage() {
+                if (this.pageDistinctColumnValues && this.chartData && this.chartMetadata.subplots) {
+                    const distinctColumn = this.chartMetadata.subplots.distinctColumn;
+                    const data = this.chartData["data"].filter((row: any) => this.pageDistinctColumnValues!.includes(row[distinctColumn]));
```

Comment:
So this function is almost an order of magnitude slower than the filtering function (about 700ms in my benchmarking) and adds considerable overhead to the scripting on this page. I wonder if the paging can be pushed back to the server a bit - each subplot is one area right? And each data point only appears on one plot? So the server could group the areas at each geographic level into pages and return that information with the data, e.g.

```
{
    "area_id": "COD",
    "area_name": "Democratic Republic of The Congo",
    "area_level": 0,
    "age_group": "Y015_049",
    "time_period": "2019",
    "quarter": "Q4",
    "plot": "anc_clients",
    "value": 4026371,
    "page:" 1
}
```

You would also need the number of pages for each area level, am still not familiar with the structure of the metadata here so not sure where that might fit, but you could also work it out once in the front end if you knew the number of plots per page and the total number of areas.

(of course this would mean that you couldn't make num plots per page configurable by the user)

A less drastic solution would be to use this proposed data structure but construct it in the front-end - replace the `allDistinctColumnValues` function with one that adds page numbers to the data points. 

Basically if we can replace this nested loop with:

```
this.chartData["data"].filter((row: any) => row["page"] == this.currentPage)
```

we'll get a speed up of 1 order of magnitude  (as long as we don't introduce any more complexity in the function that adds the page numbers.)

The below attempt is of the same order of complexity as the existing `allDistinctColumnValues` function and looks like it returns the right collection of points but for some reason its rendering them all on the same axes so I've obviously not understood something, but maybe you will understand why and be able to implement this properly!

```
  const subplots = this.chartMetadata.subplots;
  let plotCount = 0;
  let pageNum = 1;
  if (subplots && this.chartData) {
      const distinct: any[] = [];
      const subplotsPerPage = this.chartMetadata.subplots!.subplotsPerPage;
      return this.chartData["data"].map((row: any) => {
          const distinctVal = row[subplots.distinctColumn];
          if (!distinct.includes(distinctVal)) {
              distinct.push(distinctVal);
              plotCount++;
              if (plotCount == subplotsPerPage * pageNum + 1) {
                  pageNum++;
                  this.totalPages = pageNum
              }
          }
          return {...row, page: pageNum}
      })
  }
  return this.chartData ? this.chartData["data"]: []
```

Reply:
I guess it could go into the kotlin - it could catch the hintr data as it comes back and assign page numbers to the rows based on configured subplots per page and geo level. Seems quite complex, would be good to avoid if possible. 

So your function here that's appending the page numbers goes into a further computed prop `pagedChartData` or something and then it does the filtering to the page number in `chartDataPage`? That would avoid the nested loop for getting the individual pages by taking the hit to calculate the page for each row when chartData changes (when filters or dataSource change). 

I don't think this will eliminate the filter select lag, unfortuntately, since that's happening in the non-paged code too, but ti should improve it.

Reply:
I was imagining the server side approach would just be done in the R, presumably at hintr level.

The function here would replace `allDistinctColumnValues`  (although yes, would need a rename to `chartDataWithPageNums` or something) and is actually of the same order of complexity as that existing function and would be executed the same number of times, so this would only be a performance speed up. Yes, still need to look into that filtering logic and see if some optimisations can be made there! But this would avoid adding even more scripting time when paging.

Reply:
(`allDistinctColumnValues` has a `includes` nested in a `reduce` and this is an `includes` nested in a `map`)